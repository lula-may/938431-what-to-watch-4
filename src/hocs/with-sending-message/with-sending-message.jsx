import React, {PureComponent} from "react";
import {connect} from "react-redux";
import PropTypes from "prop-types";
import {getActiveMovie, getCommentUploadingError, getLoadingState} from "../../reducer/data/selectors.js";
import {getAvatarUrl} from "../../reducer/user/selectors.js";
import {Operation} from "../../reducer/data/data.js";
import {movieShape} from "../../components/shapes.js";

const MESSAGE_STYLE = {
  color: `black`,
  textAlign: `center`,
};

const withSendingMessage = (Component) => {
  class WithSendingMessage extends PureComponent {
    render() {
      const {avatar, isFormBlocked, movie, onSubmit} = this.props;

      return (
        <Component
          {...this.props}
          avatar={avatar}
          isFormBlocked={isFormBlocked}
          movie={movie}
          onSubmit={onSubmit}
        >
          {this.renderMessage()}
        </Component>
      );
    }

    renderMessage() {
      const {hasError, isFormBlocked} = this.props;
      if (isFormBlocked) {
        return (
          <p style={MESSAGE_STYLE}>Sending your review...</p>
        );
      }
      if (hasError) {
        return (
          <div style={MESSAGE_STYLE}>
            <p>Sorry, we can&apos;t post your review.</p>
            <p>Please, check the Internet connection and try again later.</p>
          </div>
        );
      }
      return null;
    }
  }

  WithSendingMessage.propTypes = {
    avatar: PropTypes.string.isRequired,
    hasError: PropTypes.bool.isRequired,
    isFormBlocked: PropTypes.bool.isRequired,
    movie: PropTypes.shape(movieShape).isRequired,
    onSubmit: PropTypes.func.isRequired,
  };

  return WithSendingMessage;

};

const mapStateToProps = (state) => ({
  avatar: getAvatarUrl(state),
  movie: getActiveMovie(state),
  hasError: getCommentUploadingError(state),
  isFormBlocked: getLoadingState(state),
});

const mapDispatchToProps = (dispatch) => ({
  onSubmit: (formData) => {
    dispatch(Operation.postComment(formData));
  }
});
export {withSendingMessage};
export default (Comp) => connect(mapStateToProps, mapDispatchToProps)(withSendingMessage(Comp));
