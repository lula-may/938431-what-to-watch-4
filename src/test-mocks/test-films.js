const movieTitles = [
  `Gone With The Wind`,
  `The Godfather`,
  `Casablanca`,
  `A Streetcar Named Desire`,
  `Some Likes It Hot`,
];

const movies = movieTitles.map((title, i) => {
  return {
    title,
    id: i,
    src: `poster-${i}.jpg`
  };
});

const headerMovie = {
  title: `Jurassic Park`,
  genre: `Sci-Fi`,
  releaseYear: 1993,
  bg: `background.jpg`,
  poster: `header-poster.jpg`
};

export {movies, headerMovie};

