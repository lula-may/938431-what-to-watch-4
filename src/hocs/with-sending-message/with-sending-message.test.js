import React from "react";
import renderer from "react-test-renderer";
import PropTypes from "prop-types";
import {withSendingMessage} from "./with-sending-message.jsx";
import {testMovies} from "../../test-mocks/test-films.js";
const MockComponent = (props) => {
  const {children} = props;
  return (
    <div>
      {children}
    </div>
  );
};

MockComponent.propTypes = {
  children: PropTypes.node.isRequired,
};
const avatar = `avatar.jpg`;
const movie = testMovies[0];

const MockComponentWrapped = withSendingMessage(MockComponent);

describe(`WithSendingMessage HOC`, () => {
  it(`should render loading message inside  MockComponent`, () => {
    const tree = renderer.create(
        <MockComponentWrapped
          avatar={avatar}
          hasError={false}
          isFormBlocked={true}
          movie={movie}
          onSubmit={() => {}}
        />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`should render error message inside  MockComponent`, () => {
    const tree = renderer.create(
        <MockComponentWrapped
          avatar={avatar}
          isFormBlocked={false}
          hasError={true}
          movie={movie}
          onSubmit={() => {}}
        />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

});
