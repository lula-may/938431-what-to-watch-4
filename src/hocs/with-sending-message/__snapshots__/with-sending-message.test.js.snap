// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WithSendingMessage HOC should render error message inside  MockComponent 1`] = `
<div>
  <div
    style={
      Object {
        "color": "black",
        "textAlign": "center",
      }
    }
  >
    <p>
      Sorry, we can't post your review.
    </p>
    <p>
      Please, check the Internet connection and try again later.
    </p>
  </div>
</div>
`;

exports[`WithSendingMessage HOC should render loading message inside  MockComponent 1`] = `
<div>
  <p
    style={
      Object {
        "color": "black",
        "textAlign": "center",
      }
    }
  >
    Sending your review...
  </p>
</div>
`;
